crates/templates/src/source.rs

```diff
@@ -18,30 +18,42 @@ pub enum TemplateSource {
     ///
     /// Templates much be in a `/templates` directory under the root of the
     /// repository.
-    Git {
-        /// The URL of the Git repository from which to install templates.
-        url: Url,
-        /// The branch or tag from which to install templates; HEAD if omitted.
-        branch: Option<String>,
-    },
+    Git(GitTemplateSource),
     /// Install from a directory in the file system.
     ///
     /// Templates much be in a `/templates` directory under the specified
     /// root.
     File(PathBuf),
 }
 
+/// Settings for installing templates from a Git repository.
+#[derive(Debug)]
+pub struct GitTemplateSource {
+    /// The URL of the Git repository from which to install templates.
+    url: Url,
+    /// The branch or tag from which to install templates; inferred if omitted.
+    branch: Option<String>,
+    /// The version of the Spin client, used for branch inference.
+    // We have to pass this through because vergen is only on the root bin
+    spin_version: String,
+}
+
 impl TemplateSource {
     /// Creates a `TemplateSource` referring to the specified Git repository
     /// and branch.
-    pub fn try_from_git(git_url: impl AsRef<str>, branch: &Option<String>) -> anyhow::Result<Self> {
+    pub fn try_from_git(
+        git_url: impl AsRef<str>,
+        branch: &Option<String>,
```

Comment:
Not really relevant to this PR but from our #rust discussion: since this function requires ownership of `branch` (except in the failure case, which probably isn't worth optimizing), this doesn't need to be a ref.

Reply:
Ditto `spin_version`: https://rust-lang.github.io/api-guidelines/flexibility.html#caller-decides-where-to-copy-and-place-data-c-caller-control

Reply:
The trouble is that `branch` gets initialised from `self.branch` in the `Install` command, so we can't move it - we could clone on the caller side.  `spin_version` comes from a macro which evaluates to a `&'static str` so again would need to `to_owned` on the caller side.  I can do these, it makes no odds to me, but they feel like bothering the caller with an implementation detail (e.g. the caller doesn't care whether `TemplateSource` owns a `branch` option, or composes the git command objects and owns them, or what) - but let me know what you prefer, I'm happy either way.